File: benchmark/scenarios/matrix/runner.ts
/**
 * 矩阵测试运行器
 */
import 'dotenv/config';
import { createTTS } from 'univoice/tts';
import { MetricsCollector } from '../../metrics/collector';
import type { BenchmarkConfig, BenchmarkResult, MatrixItem } from '../../metrics/types';
import { saveSingleResult, toSingleTestResult } from '../../utils/result-writer';
import { getProviderConfigs } from './providers';
import type { AllMatrixRunOptions, ProviderMatrixConfig, ProviderMatrixRunOptions } from './types';
import {
  filterMatrixItems,
  generateMatrixScenarioName,
  printMatrixSummary,
  printProgress,
} from './utils';

/**
 * 估算音频时长（秒）
 */
function estimateAudioDuration(audioSize: number, format: string): number {
  if (audioSize === 0) return 0;

  const bitrateKbps: Record<string, number> = {
    mp3: 128,
    wav: 256,
    pcm: 256,
    ogg: 112,
    ogg_opus: 112,
    opus: 64,
  };

  const kbps = bitrateKbps[format] || 128;
  return ((audioSize / 1024) * 8) / kbps;
}

/**
 * 计算音频码率（kbps）
 */
function calculateBitrate(audioSize: number, duration: number): number {
  if (duration === 0) return 0;
  return (audioSize * 8) / duration / 1000;
}

/**
 * 运行单个矩阵测试
 */
export async function runSingleMatrixTest(
  matrixConfig: MatrixItem,
  text: { text: string },
  scenarioName: string,
  createConfig: Record<string, unknown>
): Promise<BenchmarkResult> {
  // 导入 provider 模块
  await import('univoice/tts/providers');

  // 创建 TTS 实例
  const tts = createTTS({
    ...createConfig,
    model: matrixConfig.model,
    voice: matrixConfig.voice,
    format: matrixConfig.format,
    sampleRate: matrixConfig.sampleRate,
  } as Parameters<typeof createTTS>[0]);

  const config: BenchmarkConfig = {
    inputMode: 'non-stream',
    outputMode: 'stream',
    format: matrixConfig.format,
    textLength: text.text.length,
    voice: matrixConfig.voice,
    sampleRate: matrixConfig.sampleRate,
  };

  const collector = new MetricsCollector();
  collector.setTextLength(text.text.length);
  collector.startCollecting();

  try {
    let totalAudioSize = 0;
    for await (const { audioChunk } of tts.speak(text.text, { stream: true })) {
      collector.addChunk(audioChunk);
      totalAudioSize += audioChunk.length;
    }
    collector.endCollecting();

    const result = collector.buildResult(
      tts.name,
      tts.model,
      'tts',
      scenarioName,
      config,
      'success'
    );

    // 更新质量指标
    const duration = estimateAudioDuration(totalAudioSize, matrixConfig.format);
    const bitrate = calculateBitrate(totalAudioSize, duration);
    result.quality.audioDuration = Math.round(duration * 10) / 10;
    result.quality.bitrate = Math.round(bitrate);

    // 原子化保存
    const singleResult = toSingleTestResult(result, 1);
    saveSingleResult(singleResult);

    return result;
  } catch (error) {
    collector.endCollecting();
    const result = collector.buildResult(
      matrixConfig.provider,
      matrixConfig.model,
      'tts',
      scenarioName,
      config,
      'error',
      error instanceof Error ? error.message : String(error)
    );

    // 保存失败结果
    const singleResult = toSingleTestResult(result, 1);
    saveSingleResult(singleResult);

    return result;
  }
}

/**
 * 运行单个提供商的矩阵测试场景
 */
export async function runProviderMatrixScenario(
  providerConfig: ProviderMatrixConfig,
  options?: ProviderMatrixRunOptions
): Promise<BenchmarkResult[]> {
  // 动态导入以避免循环依赖
  const { textFixtures } = await import('../../fixtures/texts');

  const results: BenchmarkResult[] = [];
  const iterations = options?.iterations || providerConfig.scenarioConfig.iterations;
  const interval = options?.interval ?? 1000;

  // 使用第一个文本进行测试
  const text = textFixtures[0];
  if (!text) {
    throw new Error('没有可用的文本测试数据');
  }

  // 应用过滤条件
  const filteredItems = filterMatrixItems(providerConfig.items, options?.filter);

  if (filteredItems.length === 0) {
    console.warn('⚠️ 没有匹配的矩阵测试项，请检查过滤条件');
    return results;
  }

  // 打印摘要
  printMatrixSummary(
    providerConfig.displayName,
    filteredItems,
    providerConfig.items.length,
    iterations,
    options?.filter
  );

  let currentTest = 0;
  const totalTests = filteredItems.length * iterations;

  console.log('开始执行矩阵测试...\n');

  for (const matrixConfig of filteredItems) {
    const scenarioName = generateMatrixScenarioName(matrixConfig);
    const createConfig = providerConfig.createConfigFactory(matrixConfig);

    for (let i = 0; i < iterations; i++) {
      currentTest++;

      const result = await runSingleMatrixTest(matrixConfig, text, scenarioName, createConfig);
      results.push(result);

      // 打印进度
      printProgress(currentTest, totalTests, scenarioName, i + 1, result);

      // 回调
      options?.onProgress?.(currentTest, totalTests, matrixConfig, result);

      // 测试间隔
      await new Promise((resolve) => setTimeout(resolve, interval));
    }
  }

  console.log(`\n矩阵测试完成! 共执行 ${currentTest} 次测试`);

  return results;
}

/**
 * 运行矩阵测试场景（支持全量或指定提供商）
 */
export async function runMatrixScenario(options?: AllMatrixRunOptions): Promise<BenchmarkResult[]> {
  const allResults: BenchmarkResult[] = [];
  const providerConfigs = getProviderConfigs();

  // 过滤提供商
  const targetProviders = options?.providers
    ? providerConfigs.filter((p) => options.providers?.includes(p.provider))
    : providerConfigs;

  if (targetProviders.length === 0) {
    console.warn('⚠️ 没有匹配的提供商');
    return allResults;
  }

  // 运行每个提供商的测试
  for (const providerConfig of targetProviders) {
    const results = await runProviderMatrixScenario(providerConfig, {
      ...options,
      provider: providerConfig.provider,
    });
    allResults.push(...results);
  }

  return allResults;
}

/**
 * 根据提供商名称获取矩阵测试配置
 */
export function getProviderMatrixConfig(providerName: string): ProviderMatrixConfig | undefined {
  const configs = getProviderConfigs();
  return configs.find((c) => c.provider === providerName);
}

File: src/benchmark/matrix/runner.rs
//! TTS Matrix Runner
//!
//! 矩阵测试运行器：遍历 MatrixItem，对每项执行 TTS 合成测试。

use uuid::Uuid;

use crate::benchmark::collector::MetricsCollector;
use crate::benchmark::matrix::filter::{filter_matrix_items, generate_matrix_scenario_name};
use crate::benchmark::matrix::providers::get_tts_provider_matrix_config;
use crate::benchmark::matrix::types::{MatrixItem, MatrixRunOptions, ProviderMatrixConfig};
use crate::benchmark::provider_factory::create_tts_provider;
use crate::benchmark::types::{BenchmarkConfig, SingleTestResult};
use crate::tts::TtsRequest;

/// 运行单个 TTS Matrix 测试
#[allow(clippy::too_many_arguments)]
pub async fn run_single_matrix_test(
    config: &ProviderMatrixConfig,
    item: &MatrixItem,
    text: &str,
    iteration: u32,
    timeout_secs: u64,
) -> SingleTestResult {
    let _ = config; // config kept for future provider-level config use
    let scenario = generate_matrix_scenario_name(item);

    let tts = match create_tts_provider(
        &item.provider,
        &item.model,
        &item.voice,
        &item.format,
        Some(item.sample_rate),
    ) {
        Ok(t) => t,
        Err(e) => {
            return error_result(
                &item.provider,
                &item.model,
                &scenario,
                iteration,
                &e.to_string(),
            );
        }
    };

    let mut collector = MetricsCollector::new();
    collector.set_text_length(text.len());
    collector.start();

    let result = tokio::time::timeout(
        std::time::Duration::from_secs(timeout_secs),
        tts.synthesize(TtsRequest {
            text: text.to_string(),
            options: None,
        }),
    )
    .await;

    match result {
        Ok(Ok(response)) => {
            collector.add_chunk(&response.audio);
            collector.stop();

            let start_time = collector.start_time_ms();
            let throughput = collector.throughput_metrics();
            let quality = collector.quality_metrics(response.audio.len(), &response.format);

            SingleTestResult {
                id: Uuid::new_v4().to_string(),
                timestamp: chrono::Utc::now().to_rfc3339(),
                provider: item.provider.clone(),
                model: item.model.clone(),
                test_type: "tts".to_string(),
                scenario,
                iteration,
                config: BenchmarkConfig {
                    input_mode: "non-stream".to_string(),
                    output_mode: "non-stream".to_string(),
                    format: item.format.clone(),
                    text_length: Some(text.len()),
                    audio_duration: None,
                    voice: Some(item.voice.clone()),
                    sample_rate: Some(item.sample_rate),
                },
                start_time,
                throughput,
                quality,
                accuracy: None,
                status: "success".to_string(),
                error: None,
            }
        }
        Ok(Err(e)) => {
            let _start_time = collector.start_time_ms();
            error_result(
                &item.provider,
                &item.model,
                &scenario,
                iteration,
                &e.to_string(),
            )
        }
        Err(_) => {
            let _start_time = collector.start_time_ms();
            error_result(&item.provider, &item.model, &scenario, iteration, "timeout")
        }
    }
}

/// 运行单个 Provider 的矩阵场景
pub async fn run_provider_matrix_scenario(
    config: &ProviderMatrixConfig,
    text: &str,
    options: &MatrixRunOptions,
) -> Vec<SingleTestResult> {
    let items = match &options.filter {
        Some(filter) => filter_matrix_items(&config.items, filter),
        None => config.items.clone(),
    };

    let total = items.len() as u32 * options.iterations;
    let mut results = Vec::with_capacity(total as usize);
    let mut current = 0u32;

    for item in &items {
        for iter in 1..=options.iterations {
            current += 1;
            let scenario = generate_matrix_scenario_name(item);
            println!(
                "    [{}/{}] {} (iter {}/{})",
                current, total, scenario, iter, options.iterations
            );

            let result =
                run_single_matrix_test(config, item, text, iter, options.timeout_secs).await;

            let status = if result.status == "success" {
                "✓"
            } else {
                "✗"
            };
            let latency = result
                .throughput
                .chunks
                .as_ref()
                .and_then(|c| c.first())
                .map(|c| c.relative_time)
                .unwrap_or(0.0);
            println!(
                "      {} first_chunk={:.0}ms status={}",
                status, latency, result.status
            );

            results.push(result);
        }
    }

    results
}

/// 运行全量 TTS 矩阵测试
pub async fn run_matrix_scenario(
    provider_name: Option<&str>,
    text: &str,
    options: &MatrixRunOptions,
) -> Vec<SingleTestResult> {
    let configs = match provider_name {
        Some(name) => {
            vec![
                get_tts_provider_matrix_config(name)
                    .unwrap_or_else(|| panic!("Unknown matrix provider: {}", name)),
            ]
        }
        None => crate::benchmark::matrix::providers::all_tts_provider_matrix_configs(),
    };

    let mut all_results = Vec::new();
    for config in &configs {
        println!(
            "  📊 Provider: {} ({}) — {} items",
            config.display_name,
            config.provider,
            config.items.len()
        );
        let results = run_provider_matrix_scenario(config, text, options).await;
        let success = results.iter().filter(|r| r.status == "success").count();
        println!(
            "  ✓ {} 完成: {}/{} 成功\n",
            config.display_name,
            success,
            results.len()
        );
        all_results.extend(results);
    }
    all_results
}

fn error_result(
    provider: &str,
    model: &str,
    scenario: &str,
    iteration: u32,
    error_msg: &str,
) -> SingleTestResult {
    SingleTestResult {
        id: Uuid::new_v4().to_string(),
        timestamp: chrono::Utc::now().to_rfc3339(),
        provider: provider.to_string(),
        model: model.to_string(),
        test_type: "tts".to_string(),
        scenario: scenario.to_string(),
        iteration,
        config: BenchmarkConfig {
            input_mode: "non-stream".to_string(),
            output_mode: "non-stream".to_string(),
            format: "unknown".to_string(),
            text_length: None,
            audio_duration: None,
            voice: None,
            sample_rate: None,
        },
        start_time: 0.0,
        throughput: crate::benchmark::types::ThroughputMetrics {
            data_rate: 0.0,
            chunk_count: 0,
            avg_chunk_size: 0.0,
            chunks: None,
        },
        quality: crate::benchmark::types::QualityMetrics {
            data_size: 0,
            text_length: None,
            audio_duration: None,
            bitrate: None,
        },
        accuracy: None,
        status: "error".to_string(),
        error: Some(error_msg.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::benchmark::matrix::providers::qwen_provider_matrix_config;

    #[tokio::test]
    async fn test_run_single_matrix_error_unknown_provider() {
        let config = qwen_provider_matrix_config();
        let item = MatrixItem {
            provider: "unknown".into(),
            model: "test".into(),
            voice: "test".into(),
            format: "pcm".into(),
            sample_rate: 16000,
        };
        let result = run_single_matrix_test(&config, &item, "hello", 1, 10).await;
        assert_eq!(result.status, "error");
        assert!(result.error.is_some());
    }
}
